File: scripts/json-to-csv.mjs
import { promises as fs } from "node:fs";
import path from "node:path";

function usage() {
  console.log("Usage: node scripts/json-to-csv.mjs <input-json> <output-csv>");
}

function isPlainObject(value) {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

function flattenObject(value, prefix = "", target = {}) {
  for (const [key, entry] of Object.entries(value)) {
    const nextKey = prefix ? `${prefix}.${key}` : key;
    if (Array.isArray(entry)) {
      if (entry.length === 0) {
        target[nextKey] = "";
        continue;
      }

      const primitivesOnly = entry.every((item) => !isPlainObject(item) && !Array.isArray(item));
      if (primitivesOnly) {
        target[nextKey] = entry.join("; ");
        continue;
      }

      const objectItems = entry.filter(isPlainObject);
      if (objectItems.length > 0) {
        // Preserve repeated child records so they can be expanded into rows.
        target.__childRows = target.__childRows || [];
        target.__childRows.push({ key: nextKey, items: objectItems });
        continue;
      }

      target[nextKey] = JSON.stringify(entry);
      continue;
    }

    if (isPlainObject(entry)) {
      flattenObject(entry, nextKey, target);
      continue;
    }

    target[nextKey] = entry;
  }

  return target;
}

function expandRows(rows) {
  const output = [];

  for (const row of rows) {
    const flattened = flattenObject(row);
    const childRows = flattened.__childRows || [];
    delete flattened.__childRows;

    if (childRows.length === 0) {
      output.push(flattened);
      continue;
    }

    let pending = [flattened];
    for (const child of childRows) {
      const nextPending = [];
      // Duplicate parent data across every child row before merging fields.
      for (const base of pending) {
        for (const item of child.items) {
          const childFlat = flattenObject(item);
          delete childFlat.__childRows;
          nextPending.push({ ...base, ...childFlat });
        }
      }
      pending = nextPending;
    }

    output.push(...pending);
  }

  return output;
}

function toCsv(rows) {
  const columns = [...new Set(rows.flatMap((row) => Object.keys(row)))];
  const escapeCell = (value) => {
    if (value === null || value === undefined) return "";
    const text = String(value);
    if (/[",\n\r]/.test(text)) {
      return `"${text.replace(/"/g, '""')}"`;
    }
    return text;
  };

  const header = columns.join(",");
  const body = rows.map((row) => columns.map((column) => escapeCell(row[column])).join(","));
  return [header, ...body].join("\n");
}

const [, , inputPath, outputPath] = process.argv;

if (!inputPath || !outputPath) {
  usage();
  process.exit(1);
}

const raw = await fs.readFile(inputPath, "utf8");
// NDJSON is treated as one JSON object per line.
const parsed = inputPath.endsWith(".ndjson")
  ? raw
      .split(/\r?\n/)
      .filter(Boolean)
      .map((line) => JSON.parse(line))
  : JSON.parse(raw);

const rows = Array.isArray(parsed) ? parsed : [parsed];
const csv = toCsv(expandRows(rows));

await fs.mkdir(path.dirname(outputPath), { recursive: true });
await fs.writeFile(outputPath, csv + "\n", "utf8");
console.log(`Wrote ${outputPath}`);

File: scripts/json-to-csv.py
from __future__ import annotations

import json
import csv
import sys
from pathlib import Path


def is_plain_object(value):
    return isinstance(value, dict)


def flatten_object(value, prefix="", target=None):
    if target is None:
        target = {}

    child_rows = target.setdefault("__childRows", [])

    for key, entry in value.items():
        next_key = f"{prefix}.{key}" if prefix else key

        if isinstance(entry, list):
            if not entry:
                target[next_key] = ""
                continue

            primitives_only = all(not isinstance(item, (dict, list)) for item in entry)
            if primitives_only:
                target[next_key] = "; ".join(map(str, entry))
                continue

            object_items = [item for item in entry if is_plain_object(item)]
            if object_items:
                # Keep repeated child records so they can be expanded later.
                child_rows.append((next_key, object_items))
                continue

            target[next_key] = json.dumps(entry)
            continue

        if isinstance(entry, dict):
            flatten_object(entry, next_key, target)
            continue

        target[next_key] = entry

    return target


def expand_rows(rows):
    output = []

    for row in rows:
        flattened = flatten_object(row)
        child_rows = flattened.pop("__childRows", [])

        if not child_rows:
            output.append(flattened)
            continue

        pending = [flattened]
        for _, items in child_rows:
            next_pending = []
            # Duplicate parent data for every nested child row.
            for base in pending:
                for item in items:
                    child_flat = flatten_object(item)
                    child_flat.pop("__childRows", None)
                    next_pending.append({**base, **child_flat})
            pending = next_pending

        output.extend(pending)

    return output


def load_input(input_path: Path):
    if input_path.suffix.lower() == ".ndjson":
        rows = []
        # NDJSON stores one JSON object per line.
        for line in input_path.read_text(encoding="utf-8").splitlines():
            if line.strip():
                rows.append(json.loads(line))
        return rows

    data = json.loads(input_path.read_text(encoding="utf-8"))
    return data if isinstance(data, list) else [data]


def main() -> int:
    if len(sys.argv) != 3:
        print("Usage: python scripts/json-to-csv.py <input-json> <output-csv>")
        return 1

    input_path = Path(sys.argv[1])
    output_path = Path(sys.argv[2])
    rows = expand_rows(load_input(input_path))

    output_path.parent.mkdir(parents=True, exist_ok=True)
    columns = []
    for row in rows:
        for column in row.keys():
            if column not in columns:
                columns.append(column)

    with output_path.open("w", encoding="utf-8", newline="") as handle:
        writer = csv.DictWriter(handle, fieldnames=columns, extrasaction="ignore")
        writer.writeheader()
        writer.writerows(rows)

    print(f"Wrote {output_path}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
